zproject/backends.py

```diff
@@ -637,6 +637,10 @@ class ZulipLDAPUserPopulator(ZulipLDAPAuthBackendBase):
     registration for organizations that use a different SSO solution
     for managing login (often via RemoteUserBackend).
     """
+    def __init__(self, realm: Optional[Realm]=None) -> None:
+        super().__init__()
+        self._realm = realm
```

Comment:
I'm concerned about this solution being fragile.  Structurally, you're setting `self._realm` when the backend class is initialized -- which if we had designed this interface correctly, would only be once at import time (I think that's how it will work in the user registration code path, at least).  I'd guess this actually breaks using the populator in the registration code path, but I'm not sure.

In any case, this approach certainly has a risk of buggy code resulting in this `_realm` value being leaked if we forget to re-initialize the class, or something.

I don't have a better idea, though; I suppose we could just add a big comment and tests for this case in both the `sync_ldap_data` code path and the registration one, but it's kinda icky.

Reply:
@timabbott What about this new approach I just pushed? This should be cleaner - we have the ``realm`` in the "LDAP user" object. The reason we can so easily replace the ``backend.populate_user(...)`` with ``ldap_user.populate_user()`` is because the former does nothing interesting other than call the latter:
(from django_auth_ldap)
```
class LDAPBackend:
    ....
    def populate_user(self, username):
        ldap_user = _LDAPUser(self, username=username)
        return ldap_user.populate_user()
```

Reply:
I think that approach works, but it's worth adding some comments explaining the reasoning, including the chain of how `django-auth-ldap` normally ends up calling `populate_user` on the `_LDAPUser` object, and the fact that we're doing this to make sure we can pass in the realm to handle cases where there's multiple matching users in different realms.

Reply:
@timabbott pushed again, with comments on the ZulipLDAPUser class and in the sync_from_ldap function.